File: mobile_app/lib/models/workout_template.dart
// Workout Template model
import 'package:intl/intl.dart';

class WorkoutTemplate {
  final String id;
  final String trainerId;
  final String name;
  final String? description;
  final String? goal;
  final String? level;
  final DateTime createdAt;
  final List<WorkoutTemplateItem> items;

  WorkoutTemplate({
    required this.id,
    required this.trainerId,
    required this.name,
    this.description,
    this.goal,
    this.level,
    required this.createdAt,
    this.items = const [],
  });

  factory WorkoutTemplate.fromJson(Map<String, dynamic> json) {
    return WorkoutTemplate(
      id: json['id'].toString(),
      trainerId: json['trainer_id'].toString(),
      name: json['name'],
      description: json['description'],
      goal: json['goal'],
      level: json['level'],
      createdAt: DateTime.parse(json['created_at']),
      items: (json['items'] as List<dynamic>?)
              ?.map((e) => WorkoutTemplateItem.fromJson(e))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'trainer_id': trainerId,
      'name': name,
      'description': description,
      'goal': goal,
      'level': level,
      'created_at': createdAt.toIso8601String(),
      'items': items.map((e) => e.toJson()).toList(),
    };
  }

  String get exerciseCount => items.length.toString();

  String get formattedDate => DateFormat('dd/MM/yyyy').format(createdAt);
}

/// Exercise prescription within a workout template
class WorkoutTemplateItem {
  final String id;
  final String templateId;
  final String exerciseId;
  final String exerciseName;
  final int sets;
  final int? repsMin;
  final int? repsMax;
  final int? durationSeconds;
  final int restSeconds;
  final String? notes;
  final String methodologyType;
  final Map<String, dynamic>? methodologyParams;
  final String? supersetId;
  final int orderIndex;

  WorkoutTemplateItem({
    required this.id,
    required this.templateId,
    required this.exerciseId,
    required this.exerciseName,
    required this.sets,
    this.repsMin,
    this.repsMax,
    this.durationSeconds,
    required this.restSeconds,
    this.notes,
    this.methodologyType = 'NORMAL',
    this.methodologyParams,
    this.supersetId,
    this.orderIndex = 0,
  });

  factory WorkoutTemplateItem.fromJson(Map<String, dynamic> json) {
    return WorkoutTemplateItem(
      id: json['id'].toString(),
      templateId: json['template_id'].toString(),
      exerciseId: json['exercise_id'].toString(),
      exerciseName: json['exercise_name'] ?? 'Unknown',
      sets: json['sets'] ?? 0,
      repsMin: json['reps_min'],
      repsMax: json['reps_max'],
      durationSeconds: json['duration_seconds'],
      restSeconds: json['rest_seconds'] ?? 60,
      notes: json['notes'],
      methodologyType: json['methodology_type'] ?? 'NORMAL',
      methodologyParams: json['methodology_params'],
      supersetId: json['superset_id']?.toString(),
      orderIndex: json['order_index'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'template_id': templateId,
      'exercise_id': exerciseId,
      'exercise_name': exerciseName,
      'sets': sets,
      'reps_min': repsMin,
      'reps_max': repsMax,
      'duration_seconds': durationSeconds,
      'rest_seconds': restSeconds,
      'notes': notes,
      'methodology_type': methodologyType,
      'methodology_params': methodologyParams,
      'superset_id': supersetId,
      'order_index': orderIndex,
    };
  }

  bool get isTimeBased => durationSeconds != null && durationSeconds! > 0;

  String get formattedDuration {
    if (durationSeconds == null) return '';
    final minutes = durationSeconds! ~/ 60;
    final seconds = durationSeconds! % 60;
    return '$minutes:${seconds.toString().padLeft(2, '0')}';
  }
}

File: backend/app/models/workout_template.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, DateTime, Integer, ForeignKey, Enum as SQLEnum, UUID as SQLUUID, JSON
from sqlalchemy.orm import relationship
from app.db.session import Base
from app.models.workout import MethodologyType


class WorkoutTemplate(Base):
    """Reusable workout template owned by a trainer."""
    __tablename__ = "workout_templates"

    id = Column(SQLUUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    trainer_id = Column(SQLUUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    name = Column(String, nullable=False)
    description = Column(String, nullable=True)
    goal = Column(String, nullable=True)  # e.g., "Hipertrofia", "Emagrecimento"
    level = Column(String, nullable=True)  # e.g., "Iniciante", "Intermediário", "Avançado"
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    trainer = relationship("User", back_populates="workout_templates")
    items = relationship(
        "WorkoutTemplateItem",
        cascade="all, delete-orphan",
        back_populates="template",
        order_by="WorkoutTemplateItem.order_index"
    )


class WorkoutTemplateItem(Base):
    """Exercise prescription within a workout template."""
    __tablename__ = "workout_template_items"

    id = Column(SQLUUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    template_id = Column(SQLUUID(as_uuid=True), ForeignKey("workout_templates.id", ondelete="CASCADE"), nullable=False)
    exercise_id = Column(SQLUUID(as_uuid=True), ForeignKey("exercise_library.id"), nullable=False)
    exercise_name = Column(String, nullable=True)  # Snapshot of exercise name at creation time
    sets = Column(Integer, nullable=False)
    reps_min = Column(Integer, nullable=True)
    reps_max = Column(Integer, nullable=True)
    duration_seconds = Column(Integer, nullable=True)  # For cardio/time-based exercises
    rest_seconds = Column(Integer, nullable=False, default=60)
    notes = Column(String, nullable=True)
    methodology_type = Column(SQLEnum(MethodologyType), nullable=False, default=MethodologyType.NORMAL)
    methodology_params = Column(JSON, nullable=True)
    superset_id = Column(SQLUUID(as_uuid=True), nullable=True)  # For grouping exercises
    order_index = Column(Integer, nullable=False, default=0)

    template = relationship("WorkoutTemplate", back_populates="items")
    exercise = relationship("ExerciseLibrary")
